crates/sqllib/src/casts.rs

```diff
@@ -82,42 +174,50 @@ macro_rules! cast_to_b_fp {
         ::paste::paste! {
             #[doc(hidden)]
             #[inline]
-            pub fn [<cast_to_b_ $type_name>]( value: $arg_type ) -> bool {
-                value != $arg_type::zero()
+            pub fn [<cast_to_b_ $type_name>]( value: $arg_type ) -> SqlResult<bool> {
+                Ok(value != $arg_type::zero())
             }
 
             #[doc(hidden)]
             #[inline]
-            pub fn [<cast_to_b_ $type_name N >]( value: Option<$arg_type> ) -> bool {
-                [<cast_to_b_ $type_name>](value.unwrap())
+            pub fn [<cast_to_b_ $type_name N >]( value: Option<$arg_type> ) -> SqlResult<bool> {
+                match value {
+                    None => Err(cast_null("bool")),
+                    Some(value) => [<cast_to_b_ $type_name>](value),
+                }
             }
 
             #[doc(hidden)]
             #[inline]
-            pub fn [<cast_to_bN_ $type_name >]( value: $arg_type ) -> Option<bool> {
-                Some([< cast_to_b_ $type_name >](value))
+            pub fn [<cast_to_bN_ $type_name >]( value: $arg_type ) -> SqlResult<Option<bool>> {
+                r2o([< cast_to_b_ $type_name >](value))
             }
 
             #[doc(hidden)]
             #[inline]
-            pub fn [<cast_to_bN_ $type_name N >]( value: Option<$arg_type> ) -> Option<bool> {
-                let value = value?;
-                [<cast_to_bN_ $type_name >](value)
+            pub fn [<cast_to_bN_ $type_name N >]( value: Option<$arg_type> ) -> SqlResult<Option<bool>> {
```

Comment:
fascinating, can't say I understand what it does maybe you can add acomment at the high level of this whole thing